libtrap/src/ifc_tcpip.c

```diff
@@ -960,410 +957,512 @@ static int client_socket_connect(void *priv, const char *dest_addr, const char *
-         if ((cl->sending_pointer == NULL) || (cl->pending_bytes == 0)) {
-            cl->sending_pointer = (void *) data;
-            cl->pending_bytes = size;
+
+         /* Check if client is still connected */
+         if (FD_ISSET(cl->sd, &disset)) {
+            res = recv(cl->sd, buffer, DEFAULT_MAX_DATA_LENGTH, 0);
+            if (res < 1) {
+               disconnect_client(c, i);
+               VERBOSE(CL_VERBOSE_LIBRARY, "Client %u disconnected", cl->id);
+               continue;
+            }
          }
 
-         send_entry_time = get_cur_timestamp();
-         result = send_all_data(c, cl->sd, &cl->sending_pointer, &cl->pending_bytes, block);
-         send_exit_time = get_cur_timestamp();
-
-         cl->timer_last = (send_exit_time - send_entry_time);
-         cl->timer_total += cl->timer_last;
-
-         switch (result) {
-         case TRAP_E_IO_ERROR:
-            server_disconnected_client(c, i);
-            failed++;
-            break;
-         case TRAP_E_OK:
-            passed++;
-            cl->client_state = CURRENT_COMPLETE;
-            break;
-         case TRAP_E_TIMEOUT:
-            failed++;
-            break;
+         /* Check if client is ready for data */
+         if (FD_ISSET(cl->sd, &set)) {
+            send_entry_time = get_cur_timestamp();
+            res = send_data(c, cl);
+            send_exit_time = get_cur_timestamp();
+
+            /* Measure how much time we spent sending to this client (in microseconds) */
+            cl->timer_last = (send_exit_time - send_entry_time);
+            cl->timer_total += cl->timer_last;
+
+            if (res != TRAP_E_OK) {
+               VERBOSE(CL_VERBOSE_OFF, "Disconnected client %d (ret val: %d)", cl->id, res);
+               disconnect_client(c, i);
+            }
          }
-         j++;
-      }
-      if (c->connected_clients == 0) {
-         /* there is no client to send to */
-         result = TRAP_E_IO_ERROR;
-         break;
       }
    }
-   pthread_mutex_unlock(&c->sending_lock);
+}
+
+static inline void insert_into_buffer(buffer_t *buffer, const void *data, uint16_t size)
+{
+   uint16_t *msize = (uint16_t *)(buffer->data + buffer->wr_index);
+   (*msize) = htons(size);
+   memcpy((void *)(msize + 1), data, size);
+   buffer->wr_index += (size + sizeof(size));
+}
+
+/**
+ * \brief Store message into buffer.
+ *
+ * \param[in] priv      pointer to module private data
+ * \param[in] data      pointer to data to write
+ * \param[in] size      size of data to write
+ * \param[in] timeout   maximum time spent waiting for the message to be stored [microseconds]
+ *
+ * \return TRAP_E_OK         Success.
+ * \return TRAP_E_TIMEOUT    Message was not stored into buffer and the attempt should be repeated.
+ * \return TRAP_E_TERMINATED Libtrap was terminated during the process.
+ */
+int tcpip_sender_send(void *priv, const void *data, uint16_t size, int timeout)
```

Comment:
The whole function utilizes quite a lot of `goto` statements which makes the code harder to understand. Consider simplifying the code by splitting it into more functions.